src/backend/riot/rateLimiter.ts

```diff
@@ -0,0 +1,159 @@
+        if (diff > 0) {
+            this.requests.push(...Array(diff).fill(now));
+        } else if (diff < 0) {
+            this.requests.sort();
+            this.requests.splice(0, -diff);
+        }
+    }
+}
+
+function parseLimits(value: string): { [seconds: string]: number } {
+    return Object.fromEntries(
+        value.split(',').map((part) => {
+            const [limit, seconds] = part.split(':');
+            return [seconds, parseInt(limit, 10)];
+        })
+    );
+}
+
+function getRateLimitStatus(headers?: Headers): RateLimitInfo[] {
+    const appRateLimit = parseLimits(headers?.get('X-App-Rate-Limit') ?? '20:1,100:120');
+    const appRateLimitCount = parseLimits(headers?.get('X-App-Rate-Limit-Count') ?? '0:1,0:120');
+    const methodRateLimit = parseLimits(headers?.get('X-Method-Rate-Limit') ?? '500:10');
+    const methodRateLimitCount = parseLimits(headers?.get('X-Method-Rate-Limit-Count') ?? '0:10');
+
+    return [
+        ...Object.entries(appRateLimit).map(([seconds, limit]) => ({
+            type: RateLimitType.App,
+            seconds: parseInt(seconds, 10),
+            limit,
+            count: appRateLimitCount[seconds],
+        })),
+        ...Object.entries(methodRateLimit).map(([seconds, limit]) => ({
+            type: RateLimitType.Method,
+            seconds: parseInt(seconds, 10),
+            limit,
+            count: methodRateLimitCount[seconds],
+        })),
+    ];
+}
+
+export class RateLimiter {
+    private appLimits!: RateLimit[];
+    private methodLimits!: Record<APIRequestMethod, RateLimit[]>;
+    private hasInitialLimits: boolean = true;
+    private retryAfter = Promise.resolve();
+    constructor(limits?: RateLimitConfig[]) {
+        this.setLimits(limits ?? INITIAL_RATE_LIMITS);
+    }
+
+    private setLimits(limits: RateLimitConfig[]) {
+        this.appLimits = limits
+            .filter(({ type }) => type === RateLimitType.App)
+            .map(({ seconds, limit }) => new RateLimit(limit, seconds));
+        const methodLimitConfigs = limits.filter(({ type }) => type === RateLimitType.Method);
+        this.methodLimits = Object.fromEntries(
+            Object.keys(APIRequestMethod).map((method) => {
+                return [
+                    method,
+                    methodLimitConfigs.map(
+                        ({ type, seconds, limit }) => new RateLimit(limit, seconds)
+                    ),
+                ];
+            })
+        ) as Record<APIRequestMethod, RateLimit[]>;
+    }
+    refreshCounts(method: APIRequestMethod, headers: Headers) {
+        const status = getRateLimitStatus(headers);
+        if (this.hasInitialLimits) {
+            this.setLimits(status);
+            this.hasInitialLimits = false;
+        }
+        for (const info of status) {
+            const limit =
+                info.type === RateLimitType.App
+                    ? this.appLimits.find((l) => l.seconds === info.seconds)
```

Comment:
[nitpick]

I assume `l` stands for limit here, but I prefer having it spelled out.

Reply:
Fixed this instance and created #6 to address the rest of the code.